File: c++/jumpif.cc
#include "jumpif.h"
#include <iostream>

using namespace std;


jumpif::jumpif(int instruccionL) : instruccionL(instruccionL), hacerSalto(false){ 
}

jumpif::~jumpif(){
}

void jumpif::ejecutar(stack<int>& pila){
    int valor = pila.top();
    pila.pop(); 
    if(valor >= 0){
       hacerSalto = true;
    }
    else{
        hacerSalto = false;
    }
}

int jumpif::modificar(int pc){
    if(hacerSalto){
        return instruccionL;
    }else{
        return pc + 1;
    }
}

std::string jumpif::mostrar(){
    return " jumpif";
}

File: java/jumpif.java
import java.util.Stack;

class jumpif extends Instruccion {

    protected int instruccionL;
    protected boolean hacerSalto;

    public jumpif(int instruccionL){
        this.instruccionL = instruccionL;
    }

    @Override
    public void ejecutar(Stack<Integer> pila){
        int valor = pila.pop(); // Desapilo y obtengo el valor de la cima de la pila
        if(valor >= 0){
            hacerSalto = true;
        }
        else {
            hacerSalto = false;
        }
        
    }
    
    @Override 
    public int incrementarPc(int pc){
        if(hacerSalto){ // Salto a la instrucción L (parámetro de la instrucción)
            return instruccionL;
        }
        else {
            return pc + 1; // Incremento el contador de programa en 1
        }
    }

    @Override
    public String mostrar(){
        return "    jumpif" + " " + instruccionL;
    }
}
